src/lib/y2packager/resolvable.rb

```diff
@@ -0,0 +1,135 @@
+    #    The missing attributes are lazy loaded, however for performance reasons
+    #    you might ask to preload the attributes right at the beginning and avoid
+    #    querying libzypp again later.
+    # @return [Array<Y2Packager::Resolvable>] Found resolvables or empty array if nothing found
+    # @see https://yast-pkg-bindings.surge.sh/ Yast::Pkg.Resolvables
+    def self.find(params, preload = [])
+      attrs = (preload + UNIQUE_ATTRIBUTES).uniq
+      Yast::Pkg.Resolvables(params, attrs).map { |r| new(r) }
+    end
+
+    #
+    # Is there any resolvable matching the requested parameters? This is similar to
+    # the .find method, just instead of a resolvable list it returns a simple Boolean.
+    #
+    # @param params [Hash<Symbol,Object>] The requested attributes
+    # @return [Boolean] `true` if any matching resolvable is found, `false` otherwise.
+    # @see .find
+    def self.any?(params)
+      Yast::Pkg.AnyResolvable(params)
+    end
+
+    #
+    # Constructor, initialize the object from a pkg-bindings resolvable hash.
+    #
+    # @param hash [Hash<Symbol,Object>] A pkg-bindings resolvable hash.
+    def initialize(hash)
+      from_hash(hash)
+    end
+
+    #
+    # Dynamically load the missing attributes from libzypp.
+    #
+    # @param method [Symbol] the method called
+    # @param args not used so far, raises ArgumentError if anything is passed
+    #
+    # @return the loaded value from libzypp
+    #
+    def method_missing(method, *args)
+      raise ArgumentError, "Method #{method} does not accept arguments" unless args.empty?
+
+      return instance_variable_get("@#{method}") if instance_variable_defined?("@#{method}")
+
+      # load a missing attribute
+      if UNIQUE_ATTRIBUTES.all? { |a| instance_variable_defined?("@#{a}") }
+        load_attribute(method)
+      else
+        raise "Missing attributes for identifying the resolvable."
+      end
+    end
+
+  private
+
+    # attributes required for identifying a resolvable
+    UNIQUE_ATTRIBUTES = [:kind, :name, :version, :arch, :source].freeze
+
+    # Load the attributes from a Hash
+    #
+    # @param hash [Hash] The resolvable Hash obtained from pkg-bindings.
+    def from_hash(hash)
+      hash.each do |k, v|
+        instance_variable_set("@#{k}", v)
+      end
+    end
+
+    #
+    # Lazy load a missing attribute.
+    #
+    # @param attr [Symbol] The required attribute to load.
+    # @return [Object] The read value.
+    def load_attribute(attr)
+      attrs = Hash[(UNIQUE_ATTRIBUTES.map { |a| [a, instance_variable_get("@#{a}")] })]
+      resolvables = Yast::Pkg.Resolvables(attrs, [attr])
+
+      # Finding more than one result is suspicious, log a warning
+      log.warn("Found several resolvables: #{resolvables.inspect}") if resolvables.size > 1
+
+      resolvable = resolvables.first
+      raise NoMethodError unless resolvable && resolvable.key?(attr.to_s)
```

Comment:
this is a bit problematic, because exception contains no details.

I suggest to return here instead nil and above in `method_missing` call `super` if nil is returned. So it raise proper one. Just compare:

```
irb(main):001:0> def a
irb(main):002:1> raise NoMethodError
irb(main):003:1> end
irb(main):006:0> Object.a
Traceback (most recent call last):
        3: from /usr/bin/irb:11:in `<main>'
        2: from (irb):6
        1: from (irb):2:in `a'
NoMethodError (NoMethodError)
irb(main):007:0> Object.b
Traceback (most recent call last):
        2: from /usr/bin/irb:11:in `<main>'
        1: from (irb):7
NoMethodError (undefined method `b' for Object:Class)
```